lib/rust/frp/src/microtasks.rs

```diff
@@ -0,0 +1,169 @@
+//! This module contains implementation of microtask scheduler that is used in implementation of
+//! some FRP nodes.
+
+use crate::prelude::*;
+use enso_callback::traits::*;
+
+use enso_callback as callback;
+use enso_web::traits::WindowOps;
+use enso_web::Closure;
+use enso_web::JsEventHandler;
+use enso_web::JsValue;
+use enso_web::Promise;
+
+
+
+/// =================
+/// === Constants ===
+/// =================
+
+/// Maximum number of times a new microtask can be scheduled from within previously scheduled
+/// microtask handler. A safety mechanism to prevent infinite loops. See [`next_tick`] for more
+/// information.
+const MAX_RESURSIVE_MICROTASKS: usize = 1000;
+
+
+
+/// =================
+/// === next_tick ===
+/// =================
+
+/// Schedules a callback for evaluation during next microtask. This is useful for scheduling tasks
+/// that should be performed right after current JS event loop iteration, but before animation
+/// frame. All tasks scheduled with this function will be performed in the same order as they were
+/// scheduled. Further microtasks are scheduled in the same render loop as long as `next_tick`
+/// is recursively called within the scheduled task, up to a safety limit defined by
+/// [`MAX_RESURSIVE_MICROTASKS`]. If the limit is reached, an error will be raised and further tasks
+/// will be scheduled after next animation frame.
+pub fn next_tick(f: impl FnOnce() + 'static) -> callback::Handle {
+    SCHEDULER.with(|schedule| schedule.add(f))
+}
+
+/// Schedules a callback for evaluation once all microtasks in `next_tick` have been performed. If
+/// that callback calls `next_tick` or `next_tick_late` again, those tasks will be scheduled after
+/// the tasks scheduled with this function, but before control is returned to the environment.
+///
+/// TODO[PG]: Use this to delay rendering logic until the microtask queue is empty.
+/// TODO - add task number after review is done
```

Comment:
I will open a task for this and link it here once we agree that this is something we actually want to be done.

Alternatives:
- accept that streams with `next_tick` can be processed after animations are already computed and rendering is done
- remove the whole thing and keep using sampler with arbtrary delay/`before_animations`, which can introduce multi-frame delay when multiple samplers are chained together